src/cli/backup/helpers_test.go

```diff
@@ -140,9 +140,11 @@ func prepM365Test(
 		recorder = strings.Builder{}
 	)
 
-	cfg, err := st.S3Config()
+	sc, err := st.StorageConfig()
 	require.NoError(t, err, clues.ToCore(err))
 
+	cfg := sc.(*storage.S3Config)
```

Comment:
Struct interpretation from interfaces should be avoided whenever possible.  The `S3Config` func is doing all the same behavior here, with better guarantees and clarity for end users.  I'd prefer we keep with that pattern.  "Accept interfaces, return structs", as the idiom goes.

Reply:
tbh I'm a bit on the fence about that one. It makes sense, until some code decides to call some non-interface function (or field) on the struct you returned when they really shouldn't be and there's coupling where you don't expect it

```go

type Bar interface {
  Print() string
}

type foo struct {}

func getFoo() *foo { return &foo{} }

func (f *foo) Print() string {
  return "hello world"
}

func (f *foo) PrettyPrint() string {
  return "hello world!"
}

func printer(b Baz) {
  fmt.Println(b.Print())
}

func main() {
  f := getFoo()
  // Now main is coupled to this specific implementation.
  fmt.Printf("got struct that says: %s\n", f.PrettyPrint())

  // printer still uses the interface though so we can change implementations.
  printer(f)
}
```

Reply:
Well, the coupling is effectively a guarantee because different service providers supply discrete data sets, and what the consumer is really after here is that data, not standardized behavior.  There is a requirement, at some level, of a priori knowledge about what data should be available or not, and no interface adequately satisfies that expectation.

Well, techincally there is an interface that satisfies it.  `GetData() map[string]string`.  Then we expect the caller to know what keys are associated with the expected provider.  But if they know the provider's keys, then they also know the provider, which means they'd also know to call `GetSpecificProviderConfig() SpecificProviderCfg`.

Reply:
>the coupling is effectively a guarantee because different service providers supply discrete data sets, and what the consumer is really after here is that data, not standardized behavior

In this particular case that's true. However, it's not true in the general case, especially for code that could be refactored to be generic code that leverages interfaces/implementations

Reply:
Generally agreed, yes.  I still think that struct extraction from an interface is generally an anti-pattern in golang.  It smells like an attempt to recreate OOP covariance, which golang explicitly avoids supporting.  That suggests the design issue lies elsewhere (ex: if the interface doesn't provide the needed behavior, then we have the wrong interface), and the solution by casting to structure is a symptom of that other issue.